static linking causes conflicting types

If `SDL_Static` is defined, several conflicting types are produced by _sdl.image.nim_ module, example:

**index.nim**

```nim
import sdl2, sdl2.image

const imgFlags: cint = IMG_INIT_PNG
if image.init(imgFlags) != imgFlags:
  raise Exception.newException(
    "SDL2 Image initialization failed, SDL error: " & $getError())
```

**No errors without static linking:**

```
nim c -d:release index.nim
```

**Compile with `--define:SDL_Static` for static linking:**

```
$ nim c -d:release --define:SDL_Static index.nim
Hint: used config file '/path/to/Nim/config/nim.cfg' [Conf]
Hint: system [Processing]
Hint: index [Processing]
Hint: sdl2 [Processing]
Hint: macros [Processing]
Hint: unsigned [Processing]
Hint: strutils [Processing]
Hint: parseutils [Processing]
Hint: math [Processing]
Hint: algorithm [Processing]
SDL2 will be statically linked. Please make sure you pass the correct linker flags (library search paths, linked libraries).
Hint: image [Processing]
CC: index
Error: execution of an external compiler program 'gcc -c  -w -O3 -fno-strict-aliasing  -I/path/to/Nim/lib -o /path/to/my_project/nimcache/index.o /path/to/my_project/nimcache/index.c' failed with exit code: 1

In file included from /path/to/my_project/nimcache/index.c:10:0:
/path/to/my_project/nimcache/index.c:209:21: error: conflicting types for ‘SDL_GetError’
 N_NIMCALL(NCSTRING, SDL_GetError)(void);
                     ^
/path/to/Nim/lib/nimbase.h:241:44: note: in definition of macro ‘N_NIMCALL’
 #  define N_NIMCALL(rettype, name) rettype name /* no modifier */
                                            ^
In file included from /usr/include/SDL2/SDL_audio.h:32:0,
                 from /usr/include/SDL2/SDL.h:36,
                 from /usr/include/SDL2/SDL_image.h:27,
                 from /path/to/my_project/nimcache/index.c:11:
/usr/include/SDL2/SDL_error.h:42:37: note: previous declaration of ‘SDL_GetError’ was here
 extern DECLSPEC const char *SDLCALL SDL_GetError(void);
                                     ^
```

---

Nim Compiler Version 0.16.1 (2017-05-07) [Linux: amd64] (latest version of devel branch)

SDL version 2.0.4+dfsg1-2ubuntu2

SDL image version 2.0.1+dfsg-2

--- a/src/sdl2.nim
+++ b/src/sdl2.nim
@@ -9,9 +9,7 @@ export unsigned, strutils.`%`
 
 {.push warning[user]: off}
 when defined(SDL_Static):
-  static: echo "SDL2 will be statically linked. Please make sure you pass the correct linker flags "&
-    "(library search paths, linked libraries)."
-  #{.passl: gorge("pkg-config --libs sdl2").}
+  static: echo "SDL_Static option is deprecated and will soon be removed. Instead please use --dynlibOverride:SDL2."
 
 else:
   when defined(windows):

--- a/src/sdl2/audio.nim
+++ b/src/sdl2/audio.nim
@@ -234,7 +234,7 @@ const
   SDL_MIX_MAXVOLUME* = 128
 
 when defined(SDL_Static):
-  {.push header: "<SDL2/SDL.h>".}
+  static: echo "SDL_Static option is deprecated and will soon be removed. Instead please use --dynlibOverride:SDL2."
 else:
   {.push callConv: cdecl, dynlib: LibName.}
 
@@ -492,7 +492,8 @@ proc closeAudioDevice*(dev: AudioDeviceID) {.
 # Ends C function definitions when using C++
 
 # vi: set ts=4 sw=4 expandtab:
-{.pop.}
+when not defined(SDL_Static):
+  {.pop.}
 
 {.deprecated: [TAudioCVT: AudioCVT].}
 {.deprecated: [TAudioCallback: AudioCallback].}

--- a/src/sdl2/gamecontroller.nim
+++ b/src/sdl2/gamecontroller.nim
@@ -67,7 +67,7 @@ type
         hat*, hatMask*: cint
 
 when defined(SDL_Static):
-  {.push header: "<SDL2/SDL.h>".}
+  static: echo "SDL_Static option is deprecated and will soon be removed. Instead please use --dynlibOverride:SDL2."
 else:
   {.push callConv: cdecl, dynlib: LibName.}
 
@@ -290,4 +290,5 @@ proc getButton* (gameController: GameControllerPtr, button: GameControllerButton
 proc close* (gameController: GameControllerPtr) {.
   importc: "SDL_GameControllerClose".}
 
-{.pop.}
+when not defined(SDL_Static):
+  {.pop.}

--- a/src/sdl2/gfx.nim
+++ b/src/sdl2/gfx.nim
@@ -36,6 +36,8 @@ when not defined(SDL_Static):
     const LibName = "libSDL2_gfx.dylib"
   else:
     const LibName = "libSDL2_gfx(|-2.0).so(|.0)"
+else:
+  static: echo "SDL_Static option is deprecated and will soon be removed. Instead please use --dynlibOverride:SDL2."
 
 const
   FPS_UPPER_LIMIT* = 200
@@ -52,9 +54,7 @@ type
 
 {.pragma: i, importc, discardable.}
 
-when defined(SDL_Static):
-  {.push header: "<SDL2/SDL2_gfxPrimitives.h>".}
-else:
+when not defined(SDL_Static):
   {.push callConv:cdecl, dynlib: LibName.}
 
 # ---- Function Prototypes
@@ -214,10 +214,6 @@ proc stringRGBA*(renderer: RendererPtr; x: int16; y: int16; s: cstring;
                  r,g,b,a: uint8): SDL_Return {.importc, discardable.}
 # Ends C function definitions when using C++
 
-when defined(SDL_Static):
-  {.pop.}
-  {.push header: "<SDL2/SDL2_rotozoom.h>".}
-
 proc rotozoomSurface*(src: SurfacePtr; angle, zoom: cdouble;
   smooth: cint): SurfacePtr {.importc.}
 proc rotozoomSurfaceXY*(src: SurfacePtr; angle, zoomX, zoomY: cdouble;
@@ -235,10 +231,6 @@ proc shrinkSurface*(src: SurfacePtr; factorx, factorY: cint): SurfacePtr {.impor
 proc rotateSurface90Degrees*(src: SurfacePtr;
   numClockwiseTurns: cint): SurfacePtr {.importc.}
 
-when defined(SDL_Static):
-  {.pop.}
-  {.push header: "<SDL2/SDL2_framerate.h>".}
-
 proc init*(manager: var FpsManager) {.importc: "SDL_initFramerate".}
 proc setFramerate*(manager: var FpsManager; rate: cint): SDL_Return {.
   importc: "SDL_setFramerate", discardable.}
@@ -246,7 +238,8 @@ proc getFramerate*(manager: var FpsManager): cint {.importc: "SDL_getFramerate".
 proc getFramecount*(manager: var FpsManager): cint {.importc: "SDL_getFramecount".}
 proc delay*(manager: var FpsManager): cint {.importc: "SDL_framerateDelay", discardable.}
 
-{.pop.}
+when not defined(SDL_Static):
+  {.pop.}
 
 from strutils import splitLines
 proc mlStringRGBA*(renderer: RendererPtr; x,y: int16, S: string, R,G,B,A: uint8, lineSpacing = 2'i16) =

--- a/src/sdl2/haptic.nim
+++ b/src/sdl2/haptic.nim
@@ -780,7 +780,7 @@ type
 
 
 when defined(SDL_Static):
-  {.push header: "<SDL2/SDL.h>".}
+  static: echo "SDL_Static option is deprecated and will soon be removed. Instead please use --dynlibOverride:SDL2."
 else:
   {.push callConv: cdecl, dynlib: LibName.}
 
@@ -1189,4 +1189,5 @@ proc rumblePlay*(haptic: HapticPtr, strength: float, length: uint32 ):cint {.imp
 
 proc rumbleStop*(haptic: HapticPtr):cint {.importc: "SDL_HapticRumbleStop".}
 
-{.pop.}
+when not defined(SDL_Static):
+  {.pop.}

--- a/src/sdl2/image.nim
+++ b/src/sdl2/image.nim
@@ -7,6 +7,8 @@ when not defined(SDL_Static):
     const LibName = "libSDL2_image.dylib"
   else:
     const LibName = "libSDL2_image(|-2.0).so(|.0)"
+else:
+  static: echo "SDL_Static option is deprecated and will soon be removed. Instead please use --dynlibOverride:SDL2."
 
 const
   IMG_INIT_JPG* = 0x00000001
@@ -14,9 +16,7 @@ const
   IMG_INIT_TIF* = 0x00000004
   IMG_INIT_WEBP* = 0x00000008
 
-when defined(SDL_Static):
-  {.push header: "<SDL2/SDL_image.h>".}
-else:
+when not defined(SDL_Static):
   {.push callConv:cdecl, dynlib: LibName.}
 
 proc linkedVersion*(): ptr SDL_version {.importc: "IMG_Linked_Version".}
@@ -83,7 +83,8 @@ proc readXPMFromArray*(xpm: cstringArray): SurfacePtr {.importc: "IMG_ReadXPMFro
 proc savePNG*(surface: SurfacePtr, file: cstring): cint {.importc: "IMG_SavePNG".}
 #"""
 
-{.pop.}
+when not defined(SDL_Static):
+  {.pop.}
 
 {.deprecated: [IMG_Init: init].}
 {.deprecated: [IMG_Linked_Version: linkedVersion].}

--- a/src/sdl2/joystick.nim
+++ b/src/sdl2/joystick.nim
@@ -61,7 +61,7 @@ type
   JoystickID* = int32
 
 when defined(SDL_Static):
-  {.push header: "<SDL2/SDL.h>".}
+  static: echo "SDL_Static option is deprecated and will soon be removed. Instead please use --dynlibOverride:SDL2."
 else:
   {.push callConv: cdecl, dynlib: LibName.}
 
@@ -267,7 +267,8 @@ proc close* (joystick: JoystickPtr) {.inline.} = joystick.joystickClose()
 # Ends C function definitions when using C++
 
 # vi: set ts=4 sw=4 expandtab:
-{.pop.}
+when not defined(SDL_Static):
+  {.pop.}
 
 {.deprecated: [PJoystick: JoystickPtr].}
 {.deprecated: [TJoystick: Joystick].}

--- a/src/sdl2/mixer.nim
+++ b/src/sdl2/mixer.nim
@@ -29,10 +29,10 @@ when not defined(SDL_Static):
     const LibName* = "libSDL2_mixer.dylib"
   else:
     const LibName* = "libSDL2_mixer(|-2.0).so(|.0)"
-
-when defined(SDL_Static):
-  {.push header: "<SDL2/SDL_mixer.h>".}
 else:
+  static: echo "SDL_Static option is deprecated and will soon be removed. Instead please use --dynlibOverride:SDL2."
+
+when not defined(SDL_Static):
   {.push callConv:cdecl, dynlib: LibName.}
 
 import sdl2, sdl2.audio
@@ -629,4 +629,5 @@ proc getChunk*(channel: cint): ptr Chunk {.importc: "Mix_GetChunk".}
 
 proc closeAudio*() {.importc: "Mix_CloseAudio".}
 
-{.pop.}
+when not defined(SDL_Static):
+  {.pop.}

--- a/src/sdl2/net.nim
+++ b/src/sdl2/net.nim
@@ -29,6 +29,8 @@ when not defined(SDL_Static):
     const LibName* = "libSDL2_net.dylib"
   else:
     const LibName* = "libSDL2_net(|-2.0).so(|.0)"
+else:
+  static: echo "SDL_Static option is deprecated and will soon be removed. Instead please use --dynlibOverride:SDL2."
 
 type
   IpAddress* = object
@@ -69,9 +71,7 @@ type
     ready*: cint
   GenericSocket* = ptr GenericSocketObj
 
-when defined(SDL_Static):
-  {.push header: "<SDL2/SDL2_net.h>".}
-else:
+when not defined(SDL_Static):
   {.push dynlib: LibName, callconv: cdecl.}
 # This function gets the version of the dynamically linked SDL_net library.
 #   it should NOT be used to fill a version structure, instead you should
@@ -277,7 +277,8 @@ proc write32* (value: uint32, dest: pointer) {.importc: "SDLNet_Write32".}
 proc read16* (src: pointer): uint16 {.importc: "SDLNet_Read16".}
 proc read32* (src: pointer): uint32 {.importc: "SDLNet_Read32".}
 
-{.pop.}
+when not defined(SDL_Static):
+  {.pop.}
 
 proc tcpAddSocket*(set: SocketSet; sock: TcpSocket): cint =
   addSocket(set, cast[GenericSocket](sock))

--- a/src/sdl2/ttf.nim
+++ b/src/sdl2/ttf.nim
@@ -7,6 +7,8 @@ when not defined(SDL_Static):
     const LibName = "libSDL2_ttf.dylib"
   else:
     const LibName = "libSDL2_ttf(|-2.0).so(|.0)"
+else:
+  static: echo "SDL_Static option is deprecated and will soon be removed. Instead please use --dynlibOverride:SDL2."
 
 import sdl2
 
@@ -35,9 +37,7 @@ type
 #//#define TTF_PATCHLEVEL		SDL_TTF_PATCHLEVEL
 ##define TTF_VERSION(X)		SDL_TTF_VERSION(X)
 
-when defined(SDL_Static):
-  {.push header: "<SDL2/SDL_ttf.h>".}
-else:
+when not defined(SDL_Static):
   {.push callConv:cdecl, dynlib:LibName.}
 
 proc ttfLinkedVersion*(): ptr SDL_version {.importc: "TTF_Linked_Version".}
@@ -199,7 +199,8 @@ proc ttfWasInit*(): bool {.importc: "TTF_WasInit".}
 proc getFontKerningSize*(font: FontPtr; prev_index, indx: cint): cint {.
   importc: "TTF_GetFontKerningSize".}
 
-{.pop.}
+when not defined(SDL_Static):
+  {.pop.}
 
 
 # For compatibility with previous versions, here are the old functions

